README.md
<p align="center"><br><img src="https://user-images.githubusercontent.com/236501/85893648-1c92e880-b7a8-11ea-926d-95355b8175c7.png" width="128" height="128" /></p>
<h3 align="center">Capacitor Media</h3>
<p align="center"><strong><code>@capacitor-community/media</code></strong></p>
<p align="center">
  Capacitor community plugin for enabling extra media capabilities
</p>

<p align="center">
  <img src="https://img.shields.io/maintenance/yes/2023?style=flat-square" />
  <a href="https://www.npmjs.com/package/@capacitor-community/media"><img src="https://img.shields.io/npm/l/@capacitor-community/media?style=flat-square" /></a>
<br>
  <a href="https://www.npmjs.com/package/@capacitor-community/media"><img src="https://img.shields.io/npm/dw/@capacitor-community/media?style=flat-square" /></a>
  <a href="https://www.npmjs.com/package/@capacitor-community/media"><img src="https://img.shields.io/npm/v/@capacitor-community/media?style=flat-square" /></a>
  <!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
<a href="#contributors"><img src="https://img.shields.io/badge/all%20contributors-10-orange?style=flat-square" /></a>
<!-- ALL-CONTRIBUTORS-BADGE:END -->

</p>

## Sponsors

<table>
  <tr>
    <td align="center">
      <a href="https://chatness.ai">
      <img src="https://cdn.chatness.ai/images/logo-white-symbol.png" width="40" />
      </a>
    </td>
    <td>
      <a href="https://chatness.ai">
      Chatness AI
      </a>
    </td>
  </tr>
</table>

## Maintainers

| Maintainer        | GitHub                                        | Social                                    |
| ----------------- | --------------------------------------------- | ----------------------------------------- |
| Nisala Kalupahana | [nkalupahana](https://github.com/nkalupahana) |                                           |
| Stewan Silva      | [stewones](https://github.com/stewones)       | [@stewones](https://twitter.com/stewones) |

## Installation

Run one of the following commands, based on what you're using:

```bash
npm install @capacitor-community/media # NPM
yarn add @capacitor-community/media # Yarn
```

This plugin is currently for Capacitor 5. Add an `@4` at the end to install for Capacitor 4.

After installing, be sure to sync by running `ionic cap sync`.

## Migrating to Capacitor 5

**A major breaking change has been made to this plugin:** Saving media on Android now takes an album identifier instead of an album name. The album identifier, like on iOS, can be obtained using `getAlbums()`. (This call will now also return empty albums made by the plugin.) To ensure people notice this significant change, the property has been renamed from `album` to `albumIdentifier`, which will need to be updated in your code. It is still optional on iOS.

## API

Unless otherwise noted, there should be full feature parity between iOS and Android. Web is not supported.

<docgen-index>

* [`getMedias(...)`](#getmedias)
* [`getAlbums()`](#getalbums)
* [`savePhoto(...)`](#savephoto)
* [`saveVideo(...)`](#savevideo)
* [`saveGif(...)`](#savegif)
* [`createAlbum(...)`](#createalbum)
* [Interfaces](#interfaces)
* [Enums](#enums)

</docgen-index>

<docgen-api>
<!--Update the source file JSDoc comments and rerun docgen to update the docs below-->

### getMedias(...)

```typescript
getMedias(options?: MediaFetchOptions | undefined) => Promise<MediaResponse>
```

Get filtered media from camera roll (pictures only currently).

[Code Examples](https://github.com/capacitor-community/media/blob/master/example/src/components/GetMedias.tsx)

| Param         | Type                                                            |
| ------------- | --------------------------------------------------------------- |
| **`options`** | <code><a href="#mediafetchoptions">MediaFetchOptions</a></code> |

**Returns:** <code>Promise&lt;<a href="#mediaresponse">MediaResponse</a>&gt;</code>

--------------------


### getAlbums()

```typescript
getAlbums() => Promise<MediaAlbumResponse>
```

Get list of albums. 

[Code Examples](https://github.com/capacitor-community/media/blob/master/example/src/components/GetAlbums.tsx)

**Returns:** <code>Promise&lt;<a href="#mediaalbumresponse">MediaAlbumResponse</a>&gt;</code>

--------------------


### savePhoto(...)

```typescript
savePhoto(options?: MediaSaveOptions | undefined) => Promise<PhotoResponse>
```

Saves a photo to the camera roll.

On Android and iOS, this supports web URLs, base64 encoded images 
(e.g. data:image/jpeg;base64,...), and local files.
On Android, all image formats supported by the user's photo viewer are supported.

On iOS, [all image formats supported by SDWebImage are supported.](https://github.com/SDWebImage/SDWebImage#supported-image-formats)
All images on iOS are converted to PNG for system compatability. 

[Code Examples](https://github.com/capacitor-community/media/blob/master/example/src/components/SaveMedia.tsx)

| Param         | Type                                                          |
| ------------- | ------------------------------------------------------------- |
| **`options`** | <code><a href="#mediasaveoptions">MediaSaveOptions</a></code> |

**Returns:** <code>Promise&lt;<a href="#photoresponse">PhotoResponse</a>&gt;</code>

--------------------


### saveVideo(...)

```typescript
saveVideo(options?: MediaSaveOptions | undefined) => Promise<PhotoResponse>
```

Saves a video to the camera roll.

On Android and iOS, this supports web URLs, base64 encoded videos 
(e.g. data:image/mp4;base64,...), and local files.
On Android, all video formats supported by the user's photo viewer are supported.
On iOS, the supported formats are based on whatever iOS supports at the time.

[Code Examples](https://github.com/capacitor-community/media/blob/master/example/src/components/SaveMedia.tsx)

| Param         | Type                                                          |
| ------------- | ------------------------------------------------------------- |
| **`options`** | <code><a href="#mediasaveoptions">MediaSaveOptions</a></code> |

**Returns:** <code>Promise&lt;<a href="#photoresponse">PhotoResponse</a>&gt;</code>

--------------------


### saveGif(...)

```typescript
saveGif(options?: MediaSaveOptions | undefined) => Promise<PhotoResponse>
```

Saves an animated GIF to the camera roll.

On Android and iOS, this supports web URLs, base64 encoded GIFs 
(e.g. data:image/gif;base64,...), and local files.
This only supports GIF files specifically.

[Code Examples](https://github.com/capacitor-community/media/blob/master/example/src/components/SaveMedia.tsx)

| Param         | Type                                                          |
| ------------- | ------------------------------------------------------------- |
| **`options`** | <code><a href="#mediasaveoptions">MediaSaveOptions</a></code> |

**Returns:** <code>Promise&lt;<a href="#photoresponse">PhotoResponse</a>&gt;</code>

--------------------


### createAlbum(...)

```typescript
createAlbum(options: MediaAlbumCreate) => Promise<void>
```

Creates an album.

[Code Examples](https://github.com/capacitor-community/media/blob/master/example/src/components/CreateDemoAlbum.tsx)

| Param         | Type                                                          |
| ------------- | ------------------------------------------------------------- |
| **`options`** | <code><a href="#mediaalbumcreate">MediaAlbumCreate</a></code> |

--------------------


### Interfaces


#### MediaResponse

| Prop         | Type                      |
| ------------ | ------------------------- |
| **`medias`** | <code>MediaAsset[]</code> |


#### MediaAsset

| Prop                  | Type                                                    | Description                                                             |
| --------------------- | ------------------------------------------------------- | ----------------------------------------------------------------------- |
| **`identifier`**      | <code>string</code>                                     | Platform-specific identifier                                            |
| **`data`**            | <code>string</code>                                     | Data for a photo asset as a base64 encoded string (JPEG only supported) |
| **`creationDate`**    | <code>string</code>                                     | ISO date string for creation date of asset                              |
| **`fullWidth`**       | <code>number</code>                                     | Full width of original asset                                            |
| **`fullHeight`**      | <code>number</code>                                     | Full height of original asset                                           |
| **`thumbnailWidth`**  | <code>number</code>                                     | Width of thumbnail preview                                              |
| **`thumbnailHeight`** | <code>number</code>                                     | Height of thumbnail preview                                             |
| **`location`**        | <code><a href="#medialocation">MediaLocation</a></code> | Location metadata for the asset                                         |


#### MediaLocation

| Prop            | Type                | Description                              |
| --------------- | ------------------- | ---------------------------------------- |
| **`latitude`**  | <code>number</code> | GPS latitude image was taken at          |
| **`longitude`** | <code>number</code> | GPS longitude image was taken at         |
| **`heading`**   | <code>number</code> | Heading of user at time image was taken  |
| **`altitude`**  | <code>number</code> | Altitude of user at time image was taken |
| **`speed`**     | <code>number</code> | Speed of user at time image was taken    |


#### MediaFetchOptions

| Prop                   | Type                                                                                                                                                                                     | Description                                                           |
| ---------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------- |
| **`quantity`**         | <code>number</code>                                                                                                                                                                      | The number of photos to fetch, sorted by last created date descending |
| **`thumbnailWidth`**   | <code>number</code>                                                                                                                                                                      | The width of thumbnail to return                                      |
| **`thumbnailHeight`**  | <code>number</code>                                                                                                                                                                      | The height of thumbnail to return                                     |
| **`thumbnailQuality`** | <code>number</code>                                                                                                                                                                      | The quality of thumbnail to return as JPEG (0-100)                    |
| **`types`**            | <code>"photos"</code>                                                                                                                                                                    | Which types of assets to return. Only photos supported currently.     |
| **`albumIdentifier`**  | <code>string</code>                                                                                                                                                                      | Which album identifier to query in (get identifier with getAlbums())  |
| **`sort`**             | <code>"mediaType" \| "mediaSubtypes" \| "sourceType" \| "pixelWidth" \| "pixelHeight" \| "creationDate" \| "modificationDate" \| "isFavorite" \| "burstIdentifier" \| MediaSort[]</code> | Sort order of returned assets by field and ascending/descending       |


#### MediaSort

| Prop            | Type                                                                                                                                                                      |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **`key`**       | <code>"mediaType" \| "mediaSubtypes" \| "sourceType" \| "pixelWidth" \| "pixelHeight" \| "creationDate" \| "modificationDate" \| "isFavorite" \| "burstIdentifier"</code> |
| **`ascending`** | <code>boolean</code>                                                                                                                                                      |


#### MediaAlbumResponse

| Prop         | Type                      |
| ------------ | ------------------------- |
| **`albums`** | <code>MediaAlbum[]</code> |


#### MediaAlbum

| Prop             | Type                                                      |
| ---------------- | --------------------------------------------------------- |
| **`identifier`** | <code>string</code>                                       |
| **`name`**       | <code>string</code>                                       |
| **`type`**       | <code><a href="#mediaalbumtype">MediaAlbumType</a></code> |


#### PhotoResponse

| Prop           | Type                | Description                |
| -------------- | ------------------- | -------------------------- |
| **`filePath`** | <code>string</code> | Available on Android only. |


#### MediaSaveOptions

| Prop                  | Type                | Description                                                                                                                                                                                                                                                                                                                                                              |
| --------------------- | ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **`path`**            | <code>string</code> | Web URL, base64 encoded URI, or local file path to save.                                                                                                                                                                                                                                                                                                                 |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [5.3.0](https://github.com/capacitor-community/media/compare/v5.2.2...v5.3.0) (2023-10-30)
- Catch `copyFile()` crash when downloading file from the Internet

## [5.2.2](https://github.com/capacitor-community/media/compare/v5.2.1...v5.2.2) (2023-08-23)
- Fix bug in `getAlbums()` with only videos in albums

## [5.2.1](https://github.com/capacitor-community/media/compare/v5.2.0...v5.2.1) (2023-08-13)
- Add custom filenames on Android
- Fix bug in `getAlbums()` with trashed items in folder

## [5.2.0](https://github.com/capacitor-community/media/compare/v5.1.0...v5.2.0) (2023-08-13)
- Allow add-only photo permissions on iOS 14+

## [5.1.0](https://github.com/capacitor-community/media/compare/v5.0.1...v5.1.0) (2023-08-04)
- Keep EXIF data when saving images with `savePhoto()`

## [5.0.1](https://github.com/capacitor-community/media/compare/v5.0.0...v5.0.1) (2023-06-01)
- Fixes permissions on Android 13 (API 33+)

## [5.0.0](https://github.com/capacitor-community/media/compare/4.2.0...v5.0.0) (2023-05-11)
- Updates to Capacitor 5
- Returns album identifiers on Android
- BREAKING: Requires album identifiers instead of album name to add media to albums on Android
- Returns empty album names on Android

## [4.2.0](https://github.com/capacitor-community/media/compare/v4.1.0...4.2.0) (2023-03-17)

This update:
- Standardizes APIs: the iOS and Android versions of this plugin should have feature parity, bar `getMedias`.
- Updates documentation: the README is now generated via `docgen`.
- Gives code examples: almost all of the plugin's functionality can be tested using a new demo app.

## [4.1.0](https://github.com/capacitor-community/media/compare/v4.0.0-0...v4.1.0) (2023-02-20)

### Features

* **get-medias/sort:** add option to customize sort on getMedias ([#37](https://github.com/capacitor-community/media/issues/37)) ([187b253](https://github.com/capacitor-community/media/commit/187b25308415403dc11637b0baa7bc199036735f))

### [1.0.1](https://github.com/capacitor-community/media/compare/v1.0.0...v1.0.1) (2020-07-01)

### Features

- add example ([8cc960b](https://github.com/capacitor-community/media/commit/8cc960badb058d2070f313f41def4577dedcc136))

### Bug Fixes

- **android:** revamp platform. fixes [#6](https://github.com/capacitor-community/media/issues/6) ([ebdce5b](https://github.com/capacitor-community/media/commit/ebdce5bc400447fbb54c50412842ca0ae5ea6920))

## 1.0.0 (2020-06-28)

### Features

- **android:** update to capacitor v2 ([42b5620](https://github.com/capacitor-community/media/commit/42b56204e5b9aeee7b66934774a174f4b6ab8afb))
- release capacitor plugin for medias ([62a47e1](https://github.com/capacitor-community/media/commit/62a47e16dfdd8070657d206fda0e8de981ecae9e))

### Bug Fixes

- **album:** media definition ([bc07a5a](https://github.com/capacitor-community/media/commit/bc07a5aec2608c28f5b821ec1c774bac17598435))
- **android:** add app compat ([b7f8834](https://github.com/capacitor-community/media/commit/b7f8834d8ba372fa521e800383de38ace35d1321))
- **android:** enable api 29 for saveMedia ([e446af5](https://github.com/capacitor-community/media/commit/e446af568d60fdbe8809e10559a059ad94ab1ece))
- capacitor implementation ([024ce37](https://github.com/capacitor-community/media/commit/024ce37f2a14d3da676cad464af5f0d1eb5fb2a3))

android/src/main/java/com/getcapacitor/community/media/MediaPlugin.java
package com.getcapacitor.community.media;

import android.Manifest;
import android.app.DownloadManager;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.SystemClock;
import android.util.Base64;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Log;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.channels.FileChannel;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

@CapacitorPlugin(
    name = "Media",
    permissions = {
        @Permission(
            strings = { Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE },
            alias = "publicStorage"
        ),
        @Permission(
            strings = { Manifest.permission.READ_MEDIA_IMAGES, Manifest.permission.READ_MEDIA_VIDEO },
            alias = "publicStorage13Plus"
        )
    }
)
public class MediaPlugin extends Plugin {

    private static final String PERMISSION_DENIED_ERROR = "Unable to access media, user denied permission request";

    private static final int API_LEVEL_29 = 29;
    private static final int API_LEVEL_33 = 33;

    // @todo
    @PluginMethod()
    public void getMedias(PluginCall call) {
        Log.d("DEBUG LOG", "GET MEDIAS");
        if (Build.VERSION.SDK_INT >= API_LEVEL_33) {
            if (hasPermission(Manifest.permission.READ_MEDIA_IMAGES)) {
                //call.unimplemented();
                Log.d("DEBUG LOG", "HAS PERMISSION");
                _getMedias(call);
            }else{
                Log.d("DEBUG LOG", "NOT ALLOWED");
                pluginRequestPermission(Manifest.permission.READ_MEDIA_IMAGES, 1986);
            }
        } else {
            if (hasPermission(Manifest.permission.READ_EXTERNAL_STORAGE)) {
                //call.unimplemented();
                Log.d("DEBUG LOG", "HAS PERMISSION");
                _getMedias(call);
            }else{
                Log.d("DEBUG LOG", "NOT ALLOWED");
                pluginRequestPermission(Manifest.permission.READ_EXTERNAL_STORAGE, 1986);
            }
        }
    }

    // @todo Add creationDate, fullWidth, fullHeight, thumbnailWidth, thumbnailHeight, location
    private void _getMedias(PluginCall call){
        JSObject response = new JSObject();
        JSArray assets = new JSArray();

        Uri uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;

        String[] projection = {
        MediaStore.Images.Media._ID,
        MediaStore.Images.Media.DATA,
        };

        Cursor cursor = getActivity().getContentResolver().query(uri, projection, null, null, MediaStore.Images.Media._ID + " DESC");

        int columnIndexId = cursor.getColumnIndex(MediaStore.Images.Media._ID);
        int columnIndexURI = cursor.getColumnIndex(MediaStore.Images.Media.DATA);

        int start = 0;

        int quantity = call.getInt("quantity");
        if(quantity == 0) quantity = 20;

        cursor.moveToPosition(start-1); // From Index
        while (cursor.moveToNext() && cursor.getPosition() < start + quantity) { // To Index
        JSObject asset = new JSObject();
        asset.put("identifier", cursor.getString(columnIndexId));
        asset.put("data", cursor.getString(columnIndexURI)); // return filepath, have to convert Capacitor.convertFileSrc() to use in HTML & Javascript
        assets.put(asset);
        }

        cursor.close();

        response.put("medias", assets);
        call.resolve(response);
    }

    @PluginMethod
    public void getAlbums(PluginCall call) {
        Log.d("DEBUG LOG", "GET ALBUMS");
        if (isStoragePermissionGranted()) {
            Log.d("DEBUG LOG", "HAS PERMISSION");
            _getAlbums(call);
        } else {
            Log.d("DEBUG LOG", "NOT ALLOWED");
            this.bridge.saveCall(call);
            requestAllPermissions(call, "permissionCallback");
        }
    }

    @PluginMethod
    public void savePhoto(PluginCall call) {
        Log.d("DEBUG LOG", "SAVE PHOTO TO ALBUM");
        if (isStoragePermissionGranted()) {
            Log.d("DEBUG LOG", "HAS PERMISSION");
            _saveMedia(call);
        } else {
            Log.d("DEBUG LOG", "NOT ALLOWED");
            this.bridge.saveCall(call);
            requestAllPermissions(call, "permissionCallback");
            Log.d("DEBUG LOG", "___SAVE PHOTO TO ALBUM AFTER PERMISSION REQUEST");
        }
    }

    @PluginMethod
    public void saveVideo(PluginCall call) {
        Log.d("DEBUG LOG", "SAVE VIDEO TO ALBUM");
        if (isStoragePermissionGranted()) {
            Log.d("DEBUG LOG", "HAS PERMISSION");
            _saveMedia(call);
        } else {
            Log.d("DEBUG LOG", "NOT ALLOWED");
            this.bridge.saveCall(call);
            requestAllPermissions(call, "permissionCallback");
        }
    }

    @PluginMethod
    public void saveGif(PluginCall call) {
        Log.d("DEBUG LOG", "SAVE GIF TO ALBUM");
        if (isStoragePermissionGranted()) {
            Log.d("DEBUG LOG", "HAS PERMISSION");
            _saveMedia(call);
        } else {
            Log.d("DEBUG LOG", "NOT ALLOWED");
            this.bridge.saveCall(call);
            requestAllPermissions(call, "permissionCallback");
        }
    }

    @PluginMethod
    public void createAlbum(PluginCall call) {
        Log.d("DEBUG LOG", "CREATE ALBUM");
        if (isStoragePermissionGranted()) {
            Log.d("DEBUG LOG", "HAS PERMISSION");
            _createAlbum(call);
        } else {
            Log.d("DEBUG LOG", "NOT ALLOWED");
            this.bridge.saveCall(call);
            requestAllPermissions(call, "permissionCallback");
        }
    }

    @PermissionCallback
    private void permissionCallback(PluginCall call) {
        if (!isStoragePermissionGranted()) {
            Logger.debug(getLogTag(), "User denied storage permission");
            call.reject("Unable to do file operation, user denied permission request");
            return;
        }

        switch (call.getMethodName()) {
            case "getMedias" -> _getMedias(call);
            case "getAlbums" -> _getAlbums(call);
            case "savePhoto", "saveVideo", "saveGif" -> _saveMedia(call);
            case "createAlbum" -> _createAlbum(call);
        }
    }

    private boolean isStoragePermissionGranted() {
        String permissionSet = "publicStorage";
        if (Build.VERSION.SDK_INT >= API_LEVEL_33) {
            permissionSet = "publicStorage13Plus";
        }

        return getPermissionState(permissionSet) == PermissionState.GRANTED;
    }

    private void _getAlbums(PluginCall call) {
        Log.d("DEBUG LOG", "___GET ALBUMS");

        JSObject response = new JSObject();
        JSArray albums = new JSArray();
        Set<String> bucketIds = new HashSet<String>();
        Set<String> identifiers = new HashSet<String>();

        String[] projection = new String[] {
            MediaStore.MediaColumns.BUCKET_DISPLAY_NAME,
            MediaStore.MediaColumns.BUCKET_ID,
            MediaStore.MediaColumns.DATA
        };
        Cursor[] curs = {
            getActivity().getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, projection, null, null, null),
            getActivity().getContentResolver().query(MediaStore.Video.Media.EXTERNAL_CONTENT_URI, projection, null, null, null)
        };

        for (Cursor cur : curs) {
            while (cur.moveToNext()) {
                String albumName = cur.getString((cur.getColumnIndex(MediaStore.MediaColumns.BUCKET_DISPLAY_NAME)));
                String bucketId = cur.getString((cur.getColumnIndex(MediaStore.MediaColumns.BUCKET_ID)));

                if (!bucketIds.contains(bucketId)) {
                    String path = cur.getString((cur.getColumnIndex(MediaStore.MediaColumns.DATA)));
                    File fileForPath = new File(path);
                    JSObject album = new JSObject();

                    album.put("name", albumName);
                    album.put("identifier", fileForPath.getParent());
                    albums.put(album);

                    bucketIds.add(bucketId);
                    identifiers.add(fileForPath.getParent());
                }
            }

            cur.close();
        }

        File albumPath = new File(getAlbumPath());
        for (File sub : albumPath.listFiles()) {
            if (sub.isDirectory() && !identifiers.contains(sub.getAbsolutePath())) {
                JSObject album = new JSObject();

                album.put("name", sub.getName());
                album.put("identifier", sub.getAbsolutePath());
                identifiers.add(sub.getAbsolutePath());
                albums.put(album);
            }
        }

        response.put("albums", albums);
        Log.d("DEBUG LOG", String.valueOf(response));
        Log.d("DEBUG LOG", "___GET ALBUMS FINISHED");

        call.resolve(response);
    }

    private String getAlbumPath() {
        if (Build.VERSION.SDK_INT >= API_LEVEL_29) {
            return getContext().getExternalMediaDirs()[0].getAbsolutePath();
        } else {
            return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).getAbsolutePath();
        }
    }

    private void _saveMedia(PluginCall call) {
        Log.d("DEBUG LOG", "___SAVE MEDIA TO ALBUM");
        String inputPath = call.getString("path");
        if (inputPath == null) {
            call.reject("Input file path is required");
            return;
        }

        File inputFile;

        if (inputPath.startsWith("data:")) {
            try {
                String base64EncodedString = inputPath.substring(inputPath.indexOf(",") + 1);
                byte[] decodedBytes = Base64.decode(base64EncodedString, Base64.DEFAULT);
                inputFile = File.createTempFile(
                        "tmp",
                        "." + inputPath.split("/", 2)[1].split(";", 2)[0],
                        getContext().getCacheDir()
                );
                OutputStream os = new FileOutputStream(inputFile);
                os.write(decodedBytes);
                os.close();
            } catch (Exception e) {
                call.reject("Temporary file creation from data URL failed");
                return;
            }
        } else if (inputPath.startsWith("http://") || inputPath.startsWith("https://")) {
            DownloadManager manager = (DownloadManager) getContext().getSystemService(Context.DOWNLOAD_SERVICE);
            DownloadManager.Request request = new DownloadManager.Request(Uri.parse(inputPath));
            Uri inputUri = Uri.parse(inputPath);
            String filename = inputUri.getLastPathSegment();
            request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, filename);

            long requestID = manager.enqueue(request);
            Uri result = null;
            while (result == null) {
                SystemClock.sleep(100);
                result = manager.getUriForDownloadedFile(requestID);
            }

            final Cursor cursor = manager.query(new DownloadManager.Query().setFilterById(requestID));
            cursor.moveToFirst();
            inputPath = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
            try {
                Uri downloadsUri = Uri.parse(inputPath);
                File fileInDownloads = new File(downloadsUri.getPath());
                String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmssSSS").format(new Date());
                inputFile = copyFile(fileInDownloads, getContext().getCacheDir(), "IMG_" + timeStamp);
                fileInDownloads.delete();
                cursor.close();
            } catch (RuntimeException e) {
                call.reject("RuntimeException occurred", e);
                return;
            }
        } else {
            Uri inputUri = Uri.parse(inputPath);
            inputFile = new File(inputUri.getPath());
        }

        String album = call.getString("albumIdentifier");
        File albumDir = null;
        Log.d("SDK BUILD VERSION", String.valueOf(Build.VERSION.SDK_INT));

        if (album != null) {
            albumDir = new File(album);
        } else {
            call.error("Album identifier required");
            return;
        }

        if (!albumDir.exists() || !albumDir.isDirectory()) {
            call.error("Album identifier does not exist, use getAlbums() to get");
            return;
        }

        Log.d("ENV LOG - ALBUM DIR", String.valueOf(albumDir));

        try {
            // generate image file name using current date and time
            String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmssSSS").format(new Date());
            String fileName = call.getString("fileName", "IMG_" + timeStamp);
            File expFile = copyFile(inputFile, albumDir, fileName);
            scanPhoto(expFile);

            JSObject result = new JSObject();
            result.put("filePath", expFile.toString());
            call.resolve(result);
        } catch (RuntimeException e) {
            call.reject("RuntimeException occurred", e);
            return;
        }
    }

    private void _createAlbum(PluginCall call) {
        Log.d("DEBUG LOG", "___CREATE ALBUM");
        String folderName = call.getString("name");

        if (folderName == null) {
            call.reject("Album name must be given!");
            return;
        }

        File f = new File(getAlbumPath(), folderName);

        if (!f.exists()) {
            if (!f.mkdir()) {
                Log.d("DEBUG LOG", "___ERROR ALBUM");
                call.error("Cant create album");
            } else {
                Log.d("DEBUG LOG", "___SUCCESS ALBUM CREATED");
                call.success();
            }
        } else {
            Log.d("DEBUG LOG", "___ERROR ALBUM ALREADY EXISTS");
            call.error("Album already exists");
        }
    }

    private File copyFile(File inputFile, File albumDir, String fileName) {
        // if destination folder does not exist, create it
        if (!albumDir.exists()) {
            if (!albumDir.mkdir()) {
                throw new RuntimeException("Destination folder does not exist and cannot be created.");
            }
        }

        String absolutePath = inputFile.getAbsolutePath();
        String extension = absolutePath.substring(absolutePath.lastIndexOf("."));

        File newFile = new File(albumDir, fileName + extension);

        // Read and write image files
        FileChannel inChannel = null;
        FileChannel outChannel = null;

        try {
            inChannel = new FileInputStream(inputFile).getChannel();
        } catch (FileNotFoundException e) {
            throw new RuntimeException("Source file not found: " + inputFile + ", error: " + e.getMessage());
        }
        try {
            outChannel = new FileOutputStream(newFile).getChannel();
        } catch (FileNotFoundException e) {
            throw new RuntimeException("Copy file not found: " + newFile + ", error: " + e.getMessage());
        }

        try {
            inChannel.transferTo(0, inChannel.size(), outChannel);
        } catch (IOException e) {
            throw new RuntimeException("Error transfering file, error: " + e.getMessage());
        } finally {
            if (inChannel != null) {
                try {
                    inChannel.close();
                } catch (IOException e) {
                    Log.d("SaveImage", "Error closing input file channel: " + e.getMessage());
                    // does not harm, do nothing
                }
            }
            if (outChannel != null) {
                try {
                    outChannel.close();
                } catch (IOException e) {
                    Log.d("SaveImage", "Error closing output file channel: " + e.getMessage());
                    // does not harm, do nothing
                }
            }
        }

        return newFile;
    }

    private void scanPhoto(File imageFile) {
        Intent mediaScanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
        Uri contentUri = Uri.fromFile(imageFile);
        mediaScanIntent.setData(contentUri);
        bridge.getActivity().sendBroadcast(mediaScanIntent);
    }
}
